datalad/support/network.py

```diff
@@ -235,3 +249,503 @@ def __call__(self, url):
             return dict(mtime=r_stamp['mtime'], size=r_stamp['size'], url=url)
         finally:
             r.close()
+
+
+# TODO:  make it consistent/clear at what stage % encoding/decoding happens!
+# now it is a mix!
+
+#
+# Useful functionality in requests.models
+#  utils.requote_uri -- quote/unquote cycle to guarantee consistent appearance
+#  RequestEncodingMixin._encode_params -- Will successfully encode parameters when passed as a dict or a list of ...
+#  PreparedRequest().prepare_url(url, params) -- nicely cares about url encodings etc
+#
+
+def _guess_ri_cls(ri):
+    """Factory function which would determine which type of a ri a provided string is"""
+    # We assume that it is a URL and parse it. Depending on the result
+    # we might decide that it was something else ;)
```

Comment:
I wonder, whether this kind of logic will scale. If we come up with more special remotes or other URL magic, it might not work that way or at least require us to constantly rethink the logic of what to test first. Maybe the different RIs should have an `is_valid` method or sth., that returns true, if it can deal with that string. So the factory logic would be to test for all and if it's not ambigous, we know what to do.